File: AI SP/BFS.java
import  java.util.LinkedList;
import java.util.Queue;

class Graph {
    int vertices;
    LinkedList<Integer>[] adjList;

    @SuppressWarnings("unchecked") Graph(int vertices) {
        this.vertices = vertices;
        adjList = new LinkedList[vertices];
        for(int i=0 ; i<vertices ; i++)
            adjList[i] = new LinkedList<>();
    }

    void addEdge(int u, int v) {
        adjList[u].add(v);
    }

    void bfs(int start) {
        Queue<Integer> queue = new LinkedList<>();
        boolean[] visited = new boolean[vertices];

        visited[start] = true;
        queue.add(start);

        while(!queue.isEmpty()) {
            int current = queue.poll();
            System.out.println(current + " ");

            for(int neighbor : adjList[current]) {
                if(!visited[neighbor]) {
                    visited[neighbor] = true;
                    queue.add(neighbor);
                }
            }
        }
    }
}

public class BFS {
    public static void main(String[] args) {
        int vertices = 5;

        Graph graph = new Graph(vertices);

        graph.addEdge(0, 1);
        graph.addEdge(0, 2);
        graph.addEdge(1, 3);
        graph.addEdge(1, 4);
        graph.addEdge(2, 4);
        graph.addEdge(1, 0);
        graph.addEdge(2, 0);
        graph.addEdge(3, 1);
        graph.addEdge(4, 1);
        graph.addEdge(4, 2);

        graph.bfs(1);
    }
}

File: AI SP/bfs.py
from collections import defaultdict, deque

class Graph:
    def __init__(self):
        self.adjList = defaultdict(list) # defaultdict(list) adds an empty list to all the adjList elements

    def addEdge(self, u, v):
        self.adjList[u].append(v)
        self.adjList[v].append(u)

    def bfs(self, start):
        queue = deque()
        visited = [False] * (max(self.adjList.keys()) + 1)

        visited[start] = True
        queue.append(start)

        while queue:
            current = queue.popleft()
            print(current, end=" ")

            for neighbor in self.adjList[current]:
                if not visited[neighbor]:
                    visited[neighbor] = True
                    queue.append(neighbor)

graph = Graph()

graph.addEdge(0, 1)
graph.addEdge(0, 2)
graph.addEdge(1, 3)
graph.addEdge(1, 4)
graph.addEdge(2, 4)

graph.bfs(1)

# print(graph.adjList[5])